File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.kt
import kotlin.math.sqrt

class Solution {

    // 소수 판별 함수
    private fun isPrime(n: Int): Boolean {
        if (n < 2) return false
        for (i in 2..sqrt(n.toDouble()).toInt()) {
            if (n % i == 0) return false
        }
        return true
    }

    // DFS로 모든 숫자 조합 만들기
    private fun dfs(numbers: String, visited: BooleanArray, current: String, set: MutableSet<Int>) {
        if (current.isNotEmpty()) {
            set.add(current.toInt())
        }

        for (i in numbers.indices) {
            if (!visited[i]) {
                visited[i] = true
                dfs(numbers, visited, current + numbers[i], set)
                visited[i] = false
            }
        }
    }

    fun solution(numbers: String): Int {
        val visited = BooleanArray(numbers.length)
        val set = mutableSetOf<Int>()

        // 모든 조합 생성
        dfs(numbers, visited, "", set)

        // 소수 개수 세기
        return set.count { isPrime(it) }
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    const nums = new Set();

    // 소수 판별 함수
    function isPrime(n) {
        if (n < 2) return false;
        for (let i = 2; i * i <= n; i++) {
            if (n % i === 0) return false;
        }
        return true;
    }

    // 모든 조합 만들기 (DFS)
    function dfs(current, used) {
        if (current !== "") {
            nums.add(Number(current));
        }

        for (let i = 0; i < numbers.length; i++) {
            if (!used[i]) {
                used[i] = true;
                dfs(current + numbers[i], used);
                used[i] = false;
            }
        }
    }

    dfs("", Array(numbers.length).fill(false));

    // 소수 개수 세기
    let count = 0;
    nums.forEach(n => {
        if (isPrime(n)) count++;
    });

    return count;
}
